src/dask_awkward/lib/io/columnar.py

```diff
@@ -0,0 +1,141 @@
+            form_key_to_child_form_keys.setdefault(parent_key, []).append(child_key)
+        form_key_to_form = state["form_key_to_form"]
+        # Buffer hierarchy information
+        form_key_to_buffer_keys = state["form_key_to_buffer_keys"]
+        # Column hierarchy information
+        form_key_to_path = state["form_key_to_path"]
+
+        # Require the data of metadata buffers above shape-only requests
+        data_buffers = {
+            *report.data_touched,
+            *buffer_keys_required_to_compute_shapes(
+                parse_buffer_key,
+                report.shape_touched,
+                form_key_to_parent_form_key,
+                form_key_to_buffer_keys,
+            ),
+        }
+
+        # We can't read buffers directly, but if we encounter a metadata
+        # buffer, then we should be able to pick any child.
+        paths = set()
+        wildcard_form_key = set()
+        for buffer_key in data_buffers:
+            form_key, attribute = parse_buffer_key(buffer_key)
+            if attribute in METADATA_ATTRIBUTES:
+                wildcard_form_key.add(form_key)
+            else:
+                paths.add(form_key_to_path[form_key])
+
+        # Select the most appropriate column for each wildcard
+        for form_key in wildcard_form_key:
+            # Find (DFS) any non-empty record form in any child
+            recursive_child_forms = (
+                form_key_to_form[k]
+                for k in walk_graph_depth_first(form_key, form_key_to_child_form_keys)
+            )
+            record_form_keys_with_contents = (
+                f.form_key
+                for f in recursive_child_forms
+                if isinstance(f, ak.forms.RecordForm) and f.contents
+            )
+            # Now find the deepest of such records
+            try:
+                last_record_form_key = next(record_form_keys_with_contents)
+            except StopIteration:
+                # This is a leaf! Therefore, we read this column
+                paths.add(form_key_to_path[form_key])
+                continue
+            else:
+                # Ensure we get the "actual" last form key
+                for last_record_form_key in record_form_keys_with_contents:
+                    ...
+
+            # First see if any child is already included
+            for any_child_form_key in form_key_to_child_form_keys[last_record_form_key]:
+                any_child_path = form_key_to_path[any_child_form_key]
+                if any_child_path in paths:
+                    break
+            # Otherwise, add the last child
+            else:
+                paths.add(any_child_path)
+        return frozenset({".".join(p) for p in paths if p})
+
+    def project(
+        self: T,
+        report: TypeTracerReport,
+        state: FormStructure,
+    ) -> T:
+        if not self.use_optimization:
+            return self
+
+        return self.project_columns(self.necessary_columns(report, state))
```

Comment:
Looks like we're relying on both the JSON and Parquet reader function classes to have a `project_columns` method, should we enforce that in this mixin?:

```suggestion
        return self.project_columns(self.necessary_columns(report, state))
        
    @abc.abstractmethod
    def project_columns(columns: Sequence[str]) -> Self:
        ...
```

Reply:
I'm torn on this. We also require behavior` and `form` attributes. I think it's conventional to not add `abc` methods to mixins,

I've made a new PR that treats `self` as a protocol implementer.

Reply:
Ah yes this is a mixin 🤦 my mind was in interface mode, ignore!